cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		},
+	})
+
+	options.SetListeners([]vpcv1.LoadBalancerListenerPrototypeLoadBalancerContext{
+		{
+			Protocol: core.StringPtr("tcp"),
+			Port:     core.Int64Ptr(int64(s.APIServerPort())),
+			DefaultPool: &vpcv1.LoadBalancerPoolIdentityByName{
+				Name: core.StringPtr(fmt.Sprintf("%s-pool-%d", lb.Name, s.APIServerPort())),
+			},
+		},
+	})
+
+	for _, additionalListeners := range lb.AdditionalListeners {
+		pool := vpcv1.LoadBalancerPoolPrototype{
+			Algorithm:     core.StringPtr("round_robin"),
+			HealthMonitor: &vpcv1.LoadBalancerPoolHealthMonitorPrototype{Delay: core.Int64Ptr(5), MaxRetries: core.Int64Ptr(2), Timeout: core.Int64Ptr(2), Type: core.StringPtr("tcp")},
+			// Note: Appending port number to the name, it will be referenced to set target port while adding new pool member
+			Name:     pointer.String(fmt.Sprintf("additional-pool-%d", additionalListeners.Port)),
+			Protocol: core.StringPtr("tcp"),
+		}
+		options.Pools = append(options.Pools, pool)
+
+		listener := vpcv1.LoadBalancerListenerPrototypeLoadBalancerContext{
+			Protocol: core.StringPtr("tcp"),
+			Port:     core.Int64Ptr(additionalListeners.Port),
+			DefaultPool: &vpcv1.LoadBalancerPoolIdentityByName{
+				Name: pointer.String(fmt.Sprintf("additional-pool-%d", additionalListeners.Port)),
+			},
+		}
+		options.Listeners = append(options.Listeners, listener)
+	}
+
+	loadBalancer, _, err := s.IBMVPCClient.CreateLoadBalancer(options)
+	if err != nil {
+		return nil, err
+	}
+	lbState := infrav1beta2.VPCLoadBalancerState(*loadBalancer.ProvisioningStatus)
+	return &infrav1beta2.VPCLoadBalancerStatus{
+		ID:                loadBalancer.ID,
+		State:             lbState,
+		Hostname:          loadBalancer.Hostname,
+		ControllerCreated: pointer.Bool(true),
+	}, nil
+}
+
+func (s *PowerVSClusterScope) COSInstance() *infrav1beta2.CosInstance {
+	return s.IBMPowerVSCluster.Spec.CosInstance
+}
+
+// ReconcileCOSInstance reconcile COS bucket.
+func (s *PowerVSClusterScope) ReconcileCOSInstance() error {
+	if s.COSInstance() == nil || s.COSInstance().Name == "" {
+		return nil
+	}
+
+	// check COS service instance exist in cloud
+	cosServiceInstanceStatus, err := s.checkCOSServiceInstance()
+	if err != nil {
+		s.Error(err, "error checking cos service instance")
+		return err
+	}
+	if cosServiceInstanceStatus != nil {
+		s.SetStatus(COSInstance, infrav1beta2.ResourceReference{ID: cosServiceInstanceStatus.GUID, ControllerCreated: pointer.Bool(false)})
+	} else {
+		// create COS service instance
+		cosServiceInstanceStatus, err = s.createCOSServiceInstance()
+		if err != nil {
+			s.Error(err, "error creating cos service instance")
+			return err
+		}
+		s.SetStatus(COSInstance, infrav1beta2.ResourceReference{ID: cosServiceInstanceStatus.GUID, ControllerCreated: pointer.Bool(true)})
+	}
+
+	apiKey := os.Getenv("IBMCLOUD_API_KEY")
+	if apiKey == "" {
+		fmt.Printf("ibmcloud api key is not provided, set %s environmental variable", "IBMCLOUD_API_KEY")
+	}
+	apiKey = APIKEY
```

Comment:
how APIKEY is set?
I m seeing an To-Do remove for this. You may need to properly handle here!

Reply:
Its handled in latest changes by Prajyot